### ISSUE TITLE
Cams in Camect not appearing in plugin

### ISSUE THREAD

**Issue by DifferentComputers** [FlyingDiver/Indigo-Camect#3]
One of my cameras did not appear in the snapshot action drop down. But it used to! Switching from my camect to "any camnect and back again did not change this.

However, when I reload the plugin, the previously selected camera repopulated in the action.

So some sort of loss of communication that doesn't throw errors? Trying to use that action before reloading the plugin did nothing, but also did not report an error.

**Comment by FlyingDiver** [comment]
The only way I can think of that happening is if the plugin or camect device was restarted while one of the cameras was offline and not reported by the Camect hub.  The dict of available cameras is only populated when the Camect device starts up.

Is this in any way repeatable?

**Comment by DifferentComputers** [comment]
I'll watch for repeats. While I suspect it has happened more than once, today is the first time I've investigated in enough detail to notice that the action group was missing the camera.

**Comment by DifferentComputers** [comment]
Here's what just tipped me off to look for the issue:

`Camect Error                    Error in plugin execution ExecuteAction:

Traceback (most recent call last):
  File "plugin.py", line 241, in snapshotCameraCommand
KeyError: (u'ba58a4b07d4aceae34ae',)

   Camect Error                    Error in plugin execution ExecuteAction:

Traceback (most recent call last):
  File "plugin.py", line 241, in snapshotCameraCommand
KeyError: (u'ba58a4b07d4aceae34ae',)
`

![Screen Shot 2021-04-26 at 2 44 53 PM](https://user-images.githubusercontent.com/36273978/116134653-26be9580-a69e-11eb-9bcd-ac0338fcbf64.png)

**Comment by FlyingDiver** [comment]
If you go into the Camect web interface (home.camect.com), click the tool icon for that camera, and then the (i) next to "Camera name:", what do you get?  Like this:
<img width="546" alt="Screen Shot 2021-04-26 at 2 52 57 PM" src="https://user-images.githubusercontent.com/72976/116135559-44402f00-a69f-11eb-9145-2bfb86950932.png">

**Comment by FlyingDiver** [comment]
It's that Id line I'm looking for.  Do you have a camera with that Id? (ba58a4b07d4aceae34ae)

**Comment by DifferentComputers** [comment]
I do. But that is NOT the camera in question. that's Front Porch, and the one I was having trouble with is Patio.
![Screen Shot 2021-04-26 at 4 31 26 PM](https://user-images.githubusercontent.com/36273978/116146867-fdf1cc80-a6ac-11eb-90ca-3720716d0d5b.png)

**Comment by DifferentComputers** [comment]
Looking further into other action groups in Indigo, I see multiple camera snapshot actions that have lost their assigned cameras, and multiple cameras missing from the drop down menu.

**Comment by FlyingDiver** [comment]
Hmmm.  I wonder if the Camect is changing the Id assigned to the camera.  That would explain the issue.  I'll have to ask.

**Comment by DifferentComputers** [comment]
Not sure how many users you have, but I'll guess I might be making more use of the snapshot action than most. I use it to pop the most recent "interesting" image onto some tablets around the house, and sometimes that's happening more than a hundred times a day.

**Comment by FlyingDiver** [comment]
I wouldn't think that the frequency would be an issue, unless the Camect itself can't handle it and is doing something weird.  The plugin shouldn't care.  Can you run debug enabled for a bit and see if you get anything useful next time you get that error?

**Comment by DifferentComputers** [comment]
I've got a lot more debug, but this seems most interesting:

`  Camect Debug                    Skipping Alert trigger 968248451, wrong camera
   Camect Debug                    Checking Alert trigger 1600269398: camectID: 970720772, cam_id: 4a74429474640721002c, object = object : (list)
`

**Comment by FlyingDiver** [comment]
The first line just means that the camera specified in trigger ID 968248451 doesn't match the camera alert it's currently processing.  As expected.

The empty object list in the next one is interesting.   That means that trigger 1600269398 has an empty list of items to match on.  Look at that one in the trigger config dialog and see what's specified. 

I just looked at that code, and having an empty list won't cause an error, it'll just never fire the trigger.

**Comment by DifferentComputers** [comment]
It's an action group and when I check the camera snapshot action, the correct camect is selected (I only have one) but the camera selection is blank. Also, I can't select the correct camera but many of my cameras are listed. About half of them.

When I manually reload the plugin, then check the snapshot action again, it is populated correctly.

**Comment by FlyingDiver** [comment]
So something is causing the internal list of cameras to get corrupted.  Hmmm.

**Comment by DifferentComputers** [comment]
Seems like this happens pretty quickly too. Nor did it happen in the first few weeks I had the system running. 

For background, I have about a dozen cameras of varying types and levels of reliability, and I haven't made any substantive changes to how Indigo interacts with Camect recently.

**Comment by FlyingDiver** [comment]
How high is the resolution on your cameras?  I had to cut my Camect back to only six cameras because it couldn't process the feeds fast enough.  I have twelve cameras in SecuritySpy.

**Comment by DifferentComputers** [comment]
Resolution also varies per camera type. I don't see a lot of stuttering or dropped frames, so I doubt that's the issue. Also, several cameras are disabled in Home mode inside Camect.

Four are 1920x1080
Six are 1280x720
One is 640x480

**Comment by FlyingDiver** [comment]
Yeah, most of mine are 4.1MP, so a lot of data.

**Comment by FlyingDiver** [comment]
And just to confirm, the string in that KeyError is the ID of the camera that was picked for that action?

**Comment by DifferentComputers** [comment]
Correct, at least in the last three examples I looked at. Different snapshots for diff cameras and all matched the ID for the camera being used.

**Comment by FlyingDiver** [comment]
OK, so it's not bogus data.  The internal list is getting corrupted or truncated.  I'll see what I can figure out.

**Comment by DifferentComputers** [comment]
` Action Group                    Patio manual Alert snapshot
   Camect Error                    Error in plugin execution ExecuteAction:

Traceback (most recent call last):
  File "plugin.py", line 241, in snapshotCameraCommand
KeyError: (u'30f311007a0820b98c80',)
`

As before, the camera referenced in KeyError is the cam that the action group called.

**Comment by FlyingDiver** [comment]
Try https://github.com/FlyingDiver/Indigo-Camect/releases/tag/0.0.6
I don't know that it fixed the problem, but I might get a better clue what's going on.

**Comment by DifferentComputers** [comment]
Just saw this and will try it now.

**Comment by DifferentComputers** [comment]
With the 0.0.6 version I now see
`Camect Error                    Error in plugin execution ExecuteAction:

Traceback (most recent call last):
  File "plugin.py", line 266, in snapshotCameraCommand
UnboundLocalError: local variable 'camect' referenced before assignment
`

**Comment by FlyingDiver** [comment]
Rats.  I'll fix that and let you know when there's another version to test.

**Comment by FlyingDiver** [comment]
Try https://github.com/FlyingDiver/Indigo-Camect/releases/tag/0.0.7
